src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -25,11 +25,247 @@
 #include "utils/timestamp.h"
 
 
+static bool CheckDeadlockForDistributedTransaction(TransactionNode *transactionNode);
+static void ResetVisitedFields(HTAB *adjacencyList);
+static List * GetSortedDistributedTransactionIdList(HTAB *adjacencyList);
+static void AssicateDistributedTransactionWithBackendProc(TransactionNode *
+														  transactionNode);
 static TransactionNode * GetOrCreateTransactionNode(HTAB *adjacencyList,
 													DistributedTransactionId *
 													transactionId);
 static uint32 DistributedTransactionIdHash(const void *key, Size keysize);
-static int DistributedTransactionIdCompare(const void *a, const void *b, Size keysize);
+static int DistributedTransactionIdCompareHash(const void *a, const void *b, Size
+											   keysize);
+static int DistributedTransactionIdCompare(const void *a, const void *b);
+
+
+PG_FUNCTION_INFO_V1(check_distributed_deadlocks);
+
+
+/*
+ * check_distributed_deadlocks is the external API for manually
+ * checking for distributed deadlocks. For the details, see
+ * CheckForDistributedDeadlocks().
+ */
+Datum
+check_distributed_deadlocks(PG_FUNCTION_ARGS)
+{
+	bool deadlockFound = CheckForDistributedDeadlocks();
+
+	return BoolGetDatum(deadlockFound);
+}
+
+
+/*
+ * CheckForDistributedDeadlocks is the entry point for detecing
+ * distributed deadlocks.
+ *
+ * In plain words, the function first builds a wait graph by
+ * adding the wait edges from the local node and then adding the
+ * remote wait edges to form a global wait graph. Later, the wait
+ * graph is converted into an adjacency list for more efficient
+ * searches. Finally, a DFS is done on the adjacency list.
+ * Finding a cycle in the graph unveils a distributed deadlock.
+ * Upon finding a deadlock, one of the participant backend is
+ * killed by sending a SIGINT signal.
+ *
+ * The function returns true if a deadlock is found. Otherwise, returns
+ * false.
+ */
+bool
+CheckForDistributedDeadlocks(void)
+{
+	WaitGraph *waitGraph = BuildGlobalWaitGraph();
+	HTAB *adjacencyList = BuildAdjacencyListsForWaitGraph(waitGraph);
+	List *sortedTransactionIds = GetSortedDistributedTransactionIdList(adjacencyList);
+	ListCell *transactionIdCell = NULL;
+
+	/*
+	 * We need to iterate on all transactions since it is not guranteed that the
+	 * transactions form a connected graph.
+	 *
+	 * We use sorted list to get (i) predictable results (ii) kill the youngest
+	 * transaction (i.e., if a DDL continues for 1 hour and deadlocks with a
+	 * SELECT continues for 10 msec, we prefer to kill the SELECT).
+	 */
```

Comment:
Makes sense, I think it'll be trivial to print the participating transactions with that.